dev/benchmarks/test_apps/stocks/lib/stock_state.dart

```diff
@@ -0,0 +1,187 @@
+class StockSettingsPath extends StockRoutePath { const StockSettingsPath(); }
+class StockSymbolPath extends StockRoutePath {
+  const StockSymbolPath(this.symbol);
+  final String symbol;
+}
+
+enum StockMode { optimistic, pessimistic }
+enum BackupMode { enabled, disabled }
+
+class StockConfiguration {
+ const StockConfiguration({
+    @required this.stockMode,
+    @required this.backupMode,
+    @required this.debugShowGrid,
+    @required this.debugShowSizes,
+    @required this.debugShowBaselines,
+    @required this.debugShowLayers,
+    @required this.debugShowPointers,
+    @required this.debugShowRainbow,
+    @required this.showPerformanceOverlay,
+    @required this.showSemanticsDebugger,
+  }) : assert(stockMode != null),
+       assert(backupMode != null),
+       assert(debugShowGrid != null),
+       assert(debugShowSizes != null),
+       assert(debugShowBaselines != null),
+       assert(debugShowLayers != null),
+       assert(debugShowPointers != null),
+       assert(debugShowRainbow != null),
+       assert(showPerformanceOverlay != null),
+       assert(showSemanticsDebugger != null);
+
+  final StockMode stockMode;
+  final BackupMode backupMode;
+  final bool debugShowGrid;
+  final bool debugShowSizes;
+  final bool debugShowBaselines;
+  final bool debugShowLayers;
+  final bool debugShowPointers;
+  final bool debugShowRainbow;
+  final bool showPerformanceOverlay;
+  final bool showSemanticsDebugger;
+
+  StockConfiguration copyWith({
+    StockMode stockMode,
+    BackupMode backupMode,
+    bool debugShowGrid,
+    bool debugShowSizes,
+    bool debugShowBaselines,
+    bool debugShowLayers,
+    bool debugShowPointers,
+    bool debugShowRainbow,
+    bool showPerformanceOverlay,
+    bool showSemanticsDebugger,
+  }) {
+    return StockConfiguration(
+      stockMode: stockMode ?? this.stockMode,
+      backupMode: backupMode ?? this.backupMode,
+      debugShowGrid: debugShowGrid ?? this.debugShowGrid,
+      debugShowSizes: debugShowSizes ?? this.debugShowSizes,
+      debugShowBaselines: debugShowBaselines ?? this.debugShowBaselines,
+      debugShowLayers: debugShowLayers ?? this.debugShowLayers,
+      debugShowPointers: debugShowPointers ?? this.debugShowPointers,
+      debugShowRainbow: debugShowRainbow ?? this.debugShowRainbow,
+      showPerformanceOverlay: showPerformanceOverlay ?? this.showPerformanceOverlay,
+      showSemanticsDebugger: showSemanticsDebugger ?? this.showSemanticsDebugger,
+    );
+  }
+}
+
+class RouterState extends ChangeNotifier {
+  RouterState();
+
+  bool get debugInitialized {
+    bool answer;
+    assert((){
+      answer = routePath != null;
+      return true;
+    }());
+    return answer;
+  }
+
+  StockRoutePath get routePath => _routePath;
+  StockRoutePath _routePath;
+  set routePath(StockRoutePath value) {
+    if (value != _routePath) {
+      _routePath = value;
+      notifyListeners();
+    }
+  }
+}
```

Comment:
would it make sense to replace the routePath getter here with something like:

```dart
   bool get showSettings => ...; // if true, we show the settings page; otherwise:
   String get symbol => ...; // if non-null, we show the symbol information page; otherwise, show home page.
```

It seems like it would be simpler (far fewer classes, for example). You'd have to have a function to set the current mode to replace the `set routePath` setter, maybe something like;

```dart
  void updateState({ bool showSettings, String symbol }) {
    assert(!showSettings || symbol == null);
    // ...set private fields that affect the getters above...
    notifyListeners;
  }
```

Reply:
Since this is app is an example for developer to use router, i think the current approach of using the route class is better. if we convolute route classes into only a few settings, thing may become out of control if developer have more route.

For example, navigating between page if we use these property will be
```dart
showSettings= true; symbol =null  // go to setting page
showSettings = false; symbol ='AAPL' // got to symbol page
showSettings= false; symbol =null // go to home page
```

vs

```dart
routePath = StockSettingPath() // go to setting page
routePath = StockSymbolPath(symbol: 'AAPL') // go to symbol page
routePath = StockHomePath() // go to Home page
```
I think the latter is more descriptive and easier to maintain

Reply:
it is almost like we are asking user to figure out
A = 1100
B = 0110
C = 0101
D = 1001

now X xor Y = 0011 , figure what is X and Y